synapseutils/sync.py

```diff
@@ -975,3 +975,49 @@ def _check_size_each_file(df):
             single_file_size = os.stat(os.path.expandvars(os.path.expanduser(file_path))).st_size
             if single_file_size == 0:
                 raise ValueError("File {} is empty, empty files cannot be uploaded to Synapse".format(file_name))
+
+
+def generate_sync_manifest(syn, directory_path, parent_id, manifest_path):
+    """Generate manifest for syncToSynapse() from a local directory."""
+    manifest_cols = ["path", "parent"]
+    manifest_rows = _walk_directory_tree(syn, directory_path, parent_id)
+    _write_manifest_data(manifest_path, manifest_cols, manifest_rows)
+
+
+def _create_folder(syn, name, parent_id):
+    """Create Synapse folder."""
+    entity = {
+        'name': name,
+        'concreteType': 'org.sagebionetworks.repo.model.Folder',
+        'parentId': parent_id
+    }
+    entity = syn.store(entity)
+    return entity
+
+
+def _walk_directory_tree(syn, path, parent_id):
+    """Replicate folder structure on Synapse and generate manifest
+    rows for files using corresponding Synapse folders as parents.
+    """
+    rows = list()
+    parents = {path: parent_id}
+    for dirpath, dirnames, filenames in os.walk(path):
+        # Replicate the folders on Synapse
+        for dirname in dirnames:
+            name = dirname
+            folder_path = os.path.join(dirpath, dirname)
+            parent_id = parents[dirpath]
+            folder = _create_folder(syn, name, parent_id)
```

Comment:
Can this be moved to `syncToSynapse()`?

Reply:
You mean to have `syncToSynapse` create a folder if it doesn't exist?

Reply:
Yeah, so the manifest generation doesn't create anything on Synapse. Maybe this would be in the form of a `--create-subfolders` parameter for `syncToSynapse()`, which would also be exposed at the CLI.

Reply:
Oh, I would agree with this.  What would we fill in for parent id then?  Should this just be tracked in a separate JIRA?

Reply:
I think this should be tackled in this PR because I agree with the comment that was made during Code Review, namely that the `synapse manifest` command shouldn't be creating anything on Synapse. 

The `parentId` column would continue to contain the synapse ID of the top-level Synapse project/folder, but with `create_subfolders=True`, `syncToSynapse()` would use the local directories in the `path` column to create subfolders on Synapse relative to the parent ID.

Reply:
Let's leave updating `syncToSynapse` out of this PR then?  My concern is that syncToSynapse doesn't normally look at the paths.  There are some edge cases we would have to think carefully about, but agreed that the creation of the manifest shouldn't create anything on Synapse.

Reply:
I'm fine with this.